Social login: Add loading state to the google social login

enejb opened the pull request:
Currently, there is no difference between a loading state and an error state on mobile.
This PR tries to add a loading state so that it is more clear that the social login is still loading.  Fixes #59023

#### Changes proposed in this Pull Request

Before:

https://user-images.githubusercontent.com/115071/145640629-6dc11f55-7e0f-4ea2-ac96-c6177e13b798.mp4

After:

https://user-images.githubusercontent.com/115071/145640635-6b388029-dd92-493e-9d5a-c2106237ffa2.mp4


#### Testing instructions
1. Load up this PR. 
2. Logout of WordPress.com
3. visit /log-in 
4. Notice that you can now see a loading state for the Google social login
5. visit /log-in/new 
6.  Notice that you can now see a loading state for the Google social login
7. visit /start/user
8.  Notice that you can now see a loading state for the Google social login

enejb committed:
Social login: Add loading state to the google social login

Currently there is no difference between a loading state an error state on mobile.
This PR tries to add a loading state so that it is more clear that the social login is still loading.  Fixes #59023

simison commented:
Nice! Thanks for quick change, it's a good intermediate towards something better.

I wonder how we could make this so that it shows visible with colors by default when opening the page, and then we show the disabled/loading later?

Maybe like:

- Show with colors (load scripts)
- Disable if needed once scripts load and we determine that they didn't load/work
- If someone clicks the button earlier, show some loader then

Feels like that would leave this better for most, and slightly worse for some but I think that's better way around.

enejb committed:
Add temporarily not loading state

That switches over to disabled or isLoading if more then 2 seconds have passed.

enejb commented:
Thanks for taking a look @simison 

I have added a new temporarilyNotLoading state that gets removed after 2 seconds and then we show the new placeholders or the disabled state depending on if we have loaded the google script yet or if the was an error. 

Another thing that I notice is that we render this /log-in page with the login `html` form.

enejb committed:
Fix React error

enejb committed:
Make the loading state only show up if the js hasn't been loaded yet.

enejb committed:
Increase the settimeout to 10

Seems to perform better

enejb commented:
@simison I updated the loading state so that you only see if if you click the "Continue with Google" button and not all the google code has been loaded yet. 

See 

https://user-images.githubusercontent.com/115071/147708287-a9793c6d-18c7-4b7b-844c-1bd144f2005e.mp4

simison commented:
Thanks @enejb ! I think the loading with full screen W is a bit jarring, and loading state (after click) in the button directly would be better. That's why I meant here in https://github.com/Automattic/wp-calypso/pull/59083#issuecomment-996280223— so instead of showing loading button state by default when just opening the page the loading state would show up only after click and only if really needed. I understand the demo is a bit extreme as it loads slow, and in many cases loading step would be faster, or not shown at all as scripts have finished loading?

@olaolusoga could we get quick design assistance here?

simison commented:
@autumnfjeld could you arrange Ganon to review this when it's ready for review? Thank you!

simison commented:
Lemme actually summarize the task a bit for design:

**Before**

When you open WordPress.com login/signup page, "login with google" stays greyed out/disabled until Google scripts finish loading in the background, and we finish checking if 3rd party cookies are disabled. On slow connections this time can be pretty long, and seem like it's entirely disabled:

https://user-images.githubusercontent.com/87168/148948775-07d77a17-2be4-4eb9-8a6e-a93b4bb2c58d.mov

On fast wifi connections it shows up faster of course:

https://user-images.githubusercontent.com/87168/148948986-c3d772f7-e41f-49de-a832-f33a7ea07f86.mov

Google button stays disabled when 3rd party cookies are disabled (so e.g. in incognito mode) and we show a popover explaining it. Except on mobile where popover doesn't show up and we should fix it. (https://github.com/Automattic/wp-calypso/issues/59022). 

<img width="929" alt="image" src="https://user-images.githubusercontent.com/87168/148947565-9e99cb94-4c8b-4114-852d-390ff22d70e9.png">

**After**

I think we should show the button enabled from the beginning, regardless if scripts are loaded or not:

<img width="386" alt="image" src="https://user-images.githubusercontent.com/87168/148949138-26242304-0348-45a2-8306-b30481a4454a.png">

When customer clicks the button, we show some kind of loading state if scripts haven't finished loading or 3rd party cookie check hasn't finished yet. _This needs design._

If all has finished loading, no loading state is needed. 

If after loading state 3rd party cookies are disabled, we show the disabled state.

donlair commented:
@simison @enejb just a heads up that I'm about to conclude the email-only registration page test described at pbxNRc-1c5-p2. TLDR: this treatment won for mobile versions (but not desktop) so the username and password fields will be removed from mobile registration soon.

enejb commented:
@simison Thanks for the feedback. 

I have to investigate this a bit to understand why the loading (W) was showing up. That is definitely not expected. It wasn't something that was added in this PR. Since on the login page, we generate the HTML form serverside the (W) should not be there. 

The part that I wanted to show in the https://github.com/Automattic/wp-calypso/pull/59083#issuecomment-1002799748  video is what happens between sec 22 - 25. 
1. The page loads and I can google sign in button doesn't look disabled. 
2. When the user clicks the button the loading state appears because the js that is supposed to pop up in the google sign-in window has not loaded yet. 
3. The window pop-up right away.

autumnfjeld commented:
> @autumnfjeld could you arrange Ganon to review this when it's ready for review? Thank you!

Of course. :) @enejb just ping us in #team-ganon or tag Ganon Team here in github and we can help with reviews.

enejb commented:
Pinging @jsnajdr and @sgomes on this PR so that you can take a look at the approach. Since they are looking at the log-in section of calypso right now. p4TIVU-9X1-p2

In my investigation, one of the reasons why the google social login is so slow to load is because it is not being requested right away but it has to wait until the calypso JS is loaded before the google sign in button js gets requested. 

But even if we load the google sign-in js for the sign-in button right away (if it is provided by the server on the initial HTML request) on a slow connection we could still be in the state inactive state which this PR is trying to fix. 

Part of the reason why this PR is not trying to load the google sign-in js via the server is to keep the scope of the PR small and the solution for me trying to so do wasn't great. Since the framework doesn't currently provide a way to include external scripts easily (or at least I wasn't able to find a way 😅 )

sgomes commented:
> But even if we load the google sign-in js for the sign-in button right away (if it is provided by the server on the initial HTML request) on a slow connection we could still be in the state inactive state which this PR is trying to fix.

The Google sign-in JS is loaded from a different origin, so there's always a chance it will be slower, since it involves potentially establishing a connection. For some classes of connections latency is a much bigger concern than bandwidth, and caches throw another layer of uncertainty on top of everything.

We could of course start the fetch earlier as you suggest, by somehow adding the script tag to the initial render, but even then you'd always need to consider the possibility that everything else will be ready before Google sign-in. And I wouldn't want to start the fetch too early anyway; hydrating the page seems more important than preparing one particular sign-in method.

As such, having a loading state for Google sign-in seems like a good first step and general solution. The alternative would be to block rendering of all of login until the button is ready, which is of course a pretty bad idea from a performance point of view 🙂

enejb commented:
Thanks for the feedback @sgomes 

> As such, having a loading state for Google sign-in seems like a good first step and general solution. The alternative would be to block rendering of all of login until the button is ready, which is of course a pretty bad idea from a performance point of view 🙂

I agree that to block render of all of the login until the button is ready is a bad idea. :) 

We already have a loading state for the Google sign-in button. 
See  

https://user-images.githubusercontent.com/115071/145640629-6dc11f55-7e0f-4ea2-ac96-c6177e13b798.mp4

What this PR does is proposed to move the loading state when the button is clicked in case the google sign-in js hasn't been loaded already. Any thoughts or concerns about this apporach? 

See between 23 - 25 sec of this Video.  
https://user-images.githubusercontent.com/115071/147708287-a9793c6d-18c7-4b7b-844c-1bd144f2005e.mp4

sgomes commented on client/components/social-buttons/google.js:
```diff
@@ -41,7 +41,10 @@ class GoogleLoginButton extends Component {
-		isDisabled: true,
+		isDisabled: false,
+		isLoading: false,
+		isInitilized: false,
```
Typo: should be `isInitialized`. Applies to the rest of the file as well.

sgomes commented on client/components/social-buttons/google.js:
```diff
@@ -106,12 +113,23 @@ class GoogleLoginButton extends Component {
+						// Make sure that handleClick Call happens on the next tick so that the popup always launches.
+						setTimeout( () => {
+							this.handleClick( this.state.isClicked );
+						}, 10 );
```
If you just want to wait for the next event loop tick, you can omit the delay parameter altogether.

sgomes commented on client/components/social-buttons/google.js:
```diff
@@ -158,9 +176,11 @@ class GoogleLoginButton extends Component {
 		// Options are documented here:
 		// https://developers.google.com/api-client-library/javascript/reference/referencedocs#gapiauth2signinoptions
-		window.gapi.auth2
+		window?.gapi.auth2
```
I think this is probably not what you intended. This only works as intended if `window` is defined but nullish, which doesn't happen neither on the server, nor on the client.

I think that it's more likely that you're checking for whether `window.gapi` (and not `window`) is defined, however. For that, you should do `window.gapi?.auth2` instead.

sgomes commented on client/components/social-buttons/google.js:
```diff
@@ -134,14 +152,14 @@ class GoogleLoginButton extends Component {
 
-		if ( this.state.isDisabled ) {
+		if ( ! this.state.isInitilized && ! this.state.isClicked ) {
+			this.setState( { isClicked: event, isLoading: true } );
```
`isClicked` should probably be renamed if it's going to hold non-boolean values. I'd rename it to something like `clickEvent` and default it to `undefined` instead of `false`.

sgomes commented on client/components/social-buttons/google.js:
```diff
@@ -41,7 +41,10 @@ class GoogleLoginButton extends Component {
 		errorRef: null,
-		isDisabled: true,
+		isDisabled: false,
+		isLoading: false,
```
`isLoading` and `isInitialized` are very similar names, and it's not immediately clear what each of them represents, so the distinction should probably be made clearer if they're both to remain.

Reading carefully through the code, however, it appears that `isLoading` is a special state that only gets used when the user clicks the button before the library has been initialized, and gets displayed until either the library loads or an error happens? If so, that condition can be derived from the rest of state and doesn't need to be its own entry. Something like `this.state.isClicked && ! this.state.initialized && ! this.state.error`, perhaps?

sgomes reviewed (commented):
Thank you, @enejb ! I should have been clearer with my previous comment; I understand there was already a loading state, but was arguing that it's a good idea to make it unambiguously different from an error 👍

As for what happens when the user clicks the button before it's ready, that's definitely a bug. I left some comments on your approach, but I defer to Team Calypso on any suggestions they might have.

enejb committed:
Fixes as suggesed by  sgomes

jsnajdr commented on client/components/social-icons/google.jsx:
```diff
@@ -19,13 +20,14 @@ export default class GoogleIcon extends PureComponent {
 
 	render() {
-		const props = omit( this.props, [ 'isDisabled' ] );
+		const props = omit( this.props, [ 'isDisabled', 'isLoading' ] );
```
While we're touching this code, it would be nice to remove the usage of `_.omit` in favor of something like:
```js
const { isDisabled, isLoading, ...props } = this.props;
```

jsnajdr commented on client/components/social-icons/google.jsx:
```diff
@@ -19,13 +20,14 @@ export default class GoogleIcon extends PureComponent {
 				className={ classNames( 'social-icons social-icons__google', {
 					'social-icons--enabled': ! this.props.isDisabled,
 					'social-icons--disabled': !! this.props.isDisabled,
+					placeholder: this.props.isLoading,
```
A better class name would be `social-icons--placeholder` because that better adheres to the BEM rules and avoids conflict with other usages of the very generic `placeholder` class name.

jsnajdr commented on client/components/social-buttons/google.js:
```diff
@@ -191,6 +206,8 @@ class GoogleLoginButton extends Component {
 			this.state.isDisabled || this.props.isFormDisabled || this.state.error
 		);
 
+		const isLoading = this.state.isClicked && ! this.state.initialized && ! this.state.error;
```
If the `loadDependency().catch()` handler reset the `clickEvent` back to `null`, then `isLoading` is equivalent to `!! this.state.clickEvent`. There is a deferred `clickEvent` if and only if the API is loading.

I would also carefully look at the "error" code path. If the API load fails, what happens to the saved `clickEvent`? Is it handled and how? What happens when I repeatedly click on the button after the API load fails?

jsnajdr commented on client/components/social-buttons/google.js:
```diff
@@ -106,6 +112,13 @@ class GoogleLoginButton extends Component {
+						// Make sure that handleClick Call happens on the next tick so that the popup always launches.
+						setTimeout( () => {
+							this.handleClick( this.state.clickEvent );
+						} );
```
Why do we need the `setTimeout`? Reading the comment it's unclear what does "popup always launches" mean. What will break if I call `this.handleClick` right away?

enejb committed:
Address issue.

enejb commented:
Thanks for all the feedback @sgomes and @jsnajdr. It is very much appreciated. 

I think this approach works pretty well but only in Chrome. 
In FF and Safari in my testing, the user has to click the button again to see the popup and be directed to login. 
This is because browsers don't like to launch popups that are not directly attached to user events such as a button click. 

I am not sure how to proceed here. 
Should we continue with this? Launch this flow and see what happends? 

Or try some other approach. cc @simison

autumnfjeld commented:
Hi @enejb  I just wanted to summarize what I think the goal is here, I got a little confused reading all the comments.  :)  This all about the Google sign-in code taking excessive time to load on slow connections and we don't want to show the grey-G because this gives an impression that the Google login functionality is disabled. 

For clarification, in all the discussions above the Google "loading state" is the grey G?

I think this is the solution you are currently proposing:
* Login screen loads
* on slow connections the Google Sign-in JS may take excessive time to load 
* even if the Google sign in isn't "ready", we will show the colored G (the normal ready state)
* if user clicks on the Google sign in and it is not ready , then we will turn the G to grey (the loading state)
  * In Safari and FF, once the Google js is loaded, the user must then click the "colored-G-contintue with Google" to proceed 
  * In Chrome the user can proceed without having to click again

Is that correct?

roo2 commented:
> I updated the loading state so that you only see if if you click the "Continue with Google" button and not all the google code has been loaded yet.
>  am not sure how to proceed here.

I think that the first approach of just having a different loading vs disabled state was better. I think it was an interesting approach to delay the popup opening but considering how it is getting blocked in FF and safari, I don't think it is worth investing in. I think the original approach solves the original problem from https://github.com/Automattic/wp-calypso/issues/59023 that we couldn't differentiate between loading and disabled

fushar commented:
> I think this approach works pretty well but only in Chrome.
> In FF and Safari in my testing, the user has to click the button again to see the popup and be directed to login.
> This is because browsers don't like to launch popups that are not directly attached to user events such as a button click.

Right, this is a **hard** limitation of some browsers, which is there to prevent malicious app from spamming users with ads.

There is a "workaround", though. If we use redirect-based flow (instead of popup-based), then this approach works in Firefox and Safari (I tested it). For example, by always returning `true` in the following function:

https://github.com/Automattic/wp-calypso/blob/fe3807894619fb505033200366f5f2c2f120e593/client/blocks/login/login-form.jsx#L222

However, I don't have any context on why we decided to use popup-based flow for Google login. (I would guess because it is less disruptive for the users.)

In any case, I also would suggest the same thing as @roo2's. Having to click the button twice in FF and Safari looks broken to me, and blindly switching the UX to use redirection instead of popup sounds radical. So, I personally prefer a middle ground, which is the original approach: simply showing loading state when the Google script thing has not fully loaded yet.

enejb commented:
@autumnfjeld Thanks for the great summary. 

I previously implemented
1. First show the button as if everything is loaded already.
2. Show a loading state if the loading is taking longer than 2 seconds. 
3. Remove the loading state once everything has loaded or we display the disabled state. 

The problem with that approach is that the 2 seconds seem arbitrary so users might end up with a flash of (loading state) content if they are on a slow connection. 

I think the current implementation is pretty unobtrusive we go from gray to full colour. I think that most users might not even notice it. 

I think the solution here might be to have a different disabled state and keep the loading state the same as we currently have.

autumnfjeld commented:
Maybe disabled the disabled stat could have the grey G and also very light text for "Continue with Google" and then a little info icon  that when clicked would explain the disabled state (is the disabled state always due to 3rd party cookies?) . **Similar to the desktop version.**   So when the user clicks the ℹ️  icon the same type of little pop up would appear, with the same explanatory text about cookies: "Please enable "third-part cookies" to connect your Google account. To learn how to do this, click here."

![image](https://user-images.githubusercontent.com/5665959/152079553-2fd1fcbf-e521-4467-be47-de913c275a24.png)

simison commented:
We can close this. :-)
